### PR TITLE
Fix agent reconnection issues and timing logic corrections

### LINKED ISSUE #354: Agent fails to retry connection after server disconnection
## Description

The Wazuh Agent hangs indefinitely when the server becomes unavailable. This issue occurs despite the expectation that the agent should wait for `retry_interval` (30 seconds by default) and attempt to reconnect. 

## Root Cause
Upon investigation, we discovered that when the agent closes a socket, the underlying Boost library throws a `boost::system::system_error` exception by definition. This exception is not handled, resulting in the termination of the coroutine responsible for managing the reconnection attempts.

## Steps to Reproduce

1. Start the agent and connect it to a server.
2. Terminate the server while the agent is connected.
3. Observe that the agent does not attempt to reconnect after the disconnection.

## Expected Behavior
The agent should wait for the configured `retry_interval` and attempt to reconnect to the server.

## Actual Behavior
The agent does not attempt to reconnect and remains idle indefinitely.

## Proposed Solution
Handle and ignore the `boost::system::system_error` exception to prevent the coroutine from terminating. This ensures that the agent can continue its reconnection logic as expected.

### PR DESCRIPTION
**Pull request by vikman90**
|Related issue|
|---|
| Closes #354|

## Description

This PR resolves issues related to the agent's behavior during server disconnections and introduces improvements in timing calculations.

## Key Changes

1. **Avoid socket closure during incomplete handshake:**
   - Modified the logic to ensure the connection socket is not prematurely closed if the TLS handshake has not been completed. This prevents unnecessary errors and ensures a cleaner reconnection process.
2. **Handle exceptions:**
   - Added handling for exceptions that may be thrown by socket and resolver methods. If such an exception occurs, it is now logged as an error instead of causing the coroutine to terminate.
3. **Corrected timing multipliers:**
   - Fixed inaccuracies in time multipliers used during reconnection logic to ensure the agent adheres to the specified `retry_interval`.
4. **Added a configuration option for the concurrency level**
   - On systems there `std::thread::hardware_concurrency` returned small values, the Agent wasn't working as expected due to the resource constraint on the number of threads.

### Motivation

These changes address critical issues where the agent would stop retrying after server disconnection due to unhandled exceptions or incorrect handling of connection states. Additionally, they ensure the timing behavior aligns with user expectations and configuration.

## Testing

- [ ] **Server Disconnection Scenarios:**
   - Tested agent behavior when the server disconnects before and after the handshake is completed.
   - Verified the agent retries reconnection according to the configured `retry_interval`.
- [ ] **Exception Handling:**
   - Simulated socket closure errors and confirmed they are logged without terminating the coroutine.
- [ ]  **Timing Validation:**
   - Ensured the agent's reconnection attempts are delayed by exactly the configured `retry_interval`.

### CODE REVIEW

**Review by TomasTurina — COMMENTED** [review]
Missing some UT.

File: src/agent/communicator/include/ihttp_socket.hpp [hunk 1]
```diff
@@ -17,7 +17,8 @@ namespace http_client
 
         /// @brief Connects the socket to the given endpoints
         /// @param endpoints The endpoints to connect to
-        virtual void Connect(const boost::asio::ip::tcp::resolver::results_type& endpoints) = 0;
+        virtual void Connect(const boost::asio::ip::tcp::resolver::results_type& endpoints,
```
Comment (TomasTurina) [inline src/agent/communicator/include/ihttp_socket.hpp:20]: Add new parameter to functions descriptions.

File: src/agent/communicator/src/http_client.cpp
Comment (TomasTurina) [inline src/agent/communicator/src/http_client.cpp:1]: Validate `socket` and `results` in `Co_PerformHttpRequest`.

File: src/agent/communicator/src/http_client.cpp [hunk 2]
```diff
@@ -125,7 +124,7 @@ namespace http_client
```
Comment (TomasTurina) [inline src/agent/communicator/src/http_client.cpp:124]: In line https://github.com/wazuh/wazuh-agent/pull/355/files#diff-301a5da8240724f147370f285f68ff252a29505ea09942939dbec96f24034596R120, make `req` `const`.

File: src/agent/communicator/src/http_client.cpp [hunk 3]
```diff
@@ -101,8 +101,7 @@ namespace http_client
                         reqParams.Endpoint,
```
Comment (TomasTurina) [inline src/agent/communicator/src/http_client.cpp:101]: Maybe the variable `code` from https://github.com/wazuh/wazuh-agent/pull/355/files#diff-301a5da8240724f147370f285f68ff252a29505ea09942939dbec96f24034596R95 can be reused as you did in `PerformHttpRequest`.

Comment (TomasTurina) [reply src/agent/communicator/src/http_client.cpp:101]: What is the difference between `boost::system::error_code` and `boost::beast::error_code`?

File: src/agent/communicator/src/http_client.cpp
Comment (TomasTurina) [reply src/agent/communicator/src/http_client.cpp:1]: The same in `PerformHttpRequestDownload`.

File: src/agent/communicator/src/http_client.cpp [hunk 4]
```diff
@@ -282,10 +311,23 @@ namespace http_client
...
-            socket->Write(req);
+
+            socket->Write(req, ec);
+
+            if (ec)
+            {
+                throw std::runtime_error("Error writing request: " + ec.message());
+            }
+
             socket->ReadToFile(res_parser, dstFilePath);
```
Comment (TomasTurina) [inline src/agent/communicator/src/http_client.cpp:331]: No error code check here?

Comment (TomasTurina) [reply src/agent/communicator/src/http_client.cpp:331]: The code of this function is pretty similar than `PerformHttpRequest`. Maybe we can create a private function for it?

File: src/agent/communicator/src/https_socket.hpp [hunk 5]
```diff
@@ -37,18 +58,40 @@ namespace http_client
...
+                                                    boost::asio::redirect_error(boost::asio::use_awaitable, code));
 
-            co_await m_ssl_socket.async_handshake(boost::asio::ssl::stream_base::client,
-                                                  boost::asio::redirect_error(boost::asio::use_awaitable, code));
+                if (code)
+                {
+                    LogDebug("boost::asio::async_connect returned error code: {} {}", code.value(), code.message());
+                }
+
+                co_await m_ssl_socket.async_handshake(boost::asio::ssl::stream_base::client,
```
Comment (TomasTurina) [inline src/agent/communicator/src/https_socket.hpp:72]: Check for `code`?

File: src/agent/communicator/src/http_client.cpp [hunk 6]
```diff
@@ -154,7 +153,7 @@ namespace http_client
                 {
                     onUnauthorized();
                 }
-                timerSleep = connectionRetry * A_SECOND_IN_MILLIS;
+                timerSleep = connectionRetry;
             }
 
             LogDebug("Response code: {}.", res.result_int());
```
Comment (TomasTurina) [inline src/agent/communicator/src/http_client.cpp:159]: Add also endpoint name.

File: src/agent/communicator/src/http_client.cpp [hunk 7]
```diff
@@ -178,19 +177,49 @@ namespace http_client
...
+            }
+
+            socket->Read(res, ec);
+
+            if (ec)
+            {
+                throw std::runtime_error("Error reading response: " + ec.message());
+            }
 
             LogDebug("Response code: {}.", res.result_int());
```
Comment (TomasTurina) [inline src/agent/communicator/src/http_client.cpp:217]: Add also endpoint name.

File: src/agent/communicator/src/http_client.cpp [hunk 8]
```diff
@@ -282,10 +311,23 @@ namespace http_client
...
+            socket->Write(req, ec);
+
+            if (ec)
+            {
+                throw std::runtime_error("Error writing request: " + ec.message());
+            }
+
             socket->ReadToFile(res_parser, dstFilePath);
 
             LogDebug("Response code: {}.", res_parser.get().result_int());
```
Comment (TomasTurina) [inline src/agent/communicator/src/http_client.cpp:333]: Add also endpoint name.

File: src/agent/communicator/src/http_client.cpp [hunk 9]
```diff
@@ -86,25 +106,40 @@ namespace http_client
 
         do
         {
-            std::time_t timerSleep = A_SECOND_IN_MILLIS;
+            if (!token || token->empty())
+            {
+                co_await WaitForTimer(timer, connectionRetry);
```
Comment (TomasTurina) [inline src/agent/communicator/src/http_client.cpp:111]: I think we should sleep a short period of time here, maybe 1s is enough.

**Review by jr0me — APPROVED** [review]
LGTM :100:

**Review by TomasTurina — APPROVED** [review]
GJ!

**Review by sdvendramini — APPROVED** [review]

**Pull request merged**